core/templates/pages/exploration-player-page/services/command-executor.service.ts

```diff
@@ -0,0 +1,210 @@
+    CONTINUE: this.continueClick,
+    ENTER_TEXT_NUMBER_UNITS: this.fillTextBox,
+    ADD_SET: this.addSet,
+    REMOVE_SET: this.removeSet,
+    ENTER_FRACTION: this.enterFraction,
+    SELECT_ITEM_BULLET: this.selectItemBullet,
+    SUBMIT: this.submit,
+  };
+  setElementsOnPage = 0;
+  hostname = '';
+  cachedOuterFrameMessage = '';
+  windowWrapperMessageService = null;
+  constructor(private windowRef: WindowRef) {
+    this.setElementsOnPage = 0;
+    this.hostname = '';
+    this.cachedOuterFrameMessage = '';
+    this.windowWrapperMessageService =
+      new WindowWrapperMessageService(windowRef);
+  }
+
+  sendParentReadyState(windowRef) {
+    this.windowWrapperMessageService.postMessageToParent(
+      'Ready to receive hostname', '*');
+  }
+
+  getOuterFrameEvents(windowRef) {
+    this.windowWrapperMessageService.addEventListener(
+      'message', (event) => {
+        var messageArray = event.data.split(' ');
+        var command = messageArray[0];
+        var message = '';
+        for (var i = 1; i < messageArray.length; i++) {
+          message = message + messageArray[i] + ' ';
+        }
+        message = message.substr(0, message.length - 1);
+        if ((command !== 'HOSTNAME' && this.hostname === '') ||
+        command === 'finishedProcessing') {
+          return;
+        } else if (command === 'CONTINUE' || command === 'SUBMIT') {
+          this.commandToFunctionMap[command](windowRef);
+        } else if (command === 'HOSTNAME') {
+          if (ServicesConstants.WHITELISTED_IFRAME_HOSTS.indexOf(message) >=
+          0) {
+            this.hostname = message;
+          }
+          if (this.cachedOuterFrameMessage !== '' &&
+          this.cachedOuterFrameMessage !== undefined) {
+            this.windowWrapperMessageService.postMessageToParent(
+              this.cachedOuterFrameMessage, this.hostname);
+          }
+        } else {
+          this.commandToFunctionMap[command](windowRef, message);
+        }
+      });
+  }
+
+  sendStateToOuterFrame(id) {
+    var stateToCommand = {
+      Continue: 'CONTINUE',
+      FractionInput: 'ENTER_FRACTION',
+      ItemSelectionInput: 'SELECT_ITEM_CHECKPOINT',
+      MultipleChoiceInput: 'SELECT_ITEM_BULLET',
+      NumberWithUnits: 'ENTER_TEXT_NUMBER_UNITS',
+      NumericExpressionInput: 'ENTER_TEXT_NUMBER_UNITS',
+      SetInput: 'SET_OPERATION',
+      TextInput: 'ENTER_TEXT_NUMBER_UNITS',
+    };
+    this.setElementsOnPage = 0;
+    if (!(id in stateToCommand)) {
+      return;
+    }
+    if (this.hostname === '') {
+      this.cachedOuterFrameMessage = stateToCommand[id];
+    } else {
+      this.windowWrapperMessageService.postMessageToParent(stateToCommand[id],
+        this.hostname);
+    }
+  }
+
+  continueClick(windowRef) {
+    try {
+      var button = windowRef.nativeWindow.document.getElementsByClassName(
+        'oppia-learner-confirm-button'
+      )[0] as HTMLElement;
+      button.click();
+    } catch {
```

Comment:
Why is this catch needed? It’s generally an antipattern to use exception handling for control flow so we should try to avoid those situations.